File: FWFramework/Component/Foundation/Extension/NSFileManager+FWFramework.m
/*!
 @header     NSFileManager+FWFramework.m
 @indexgroup FWFramework
 @brief      NSFileManager+FWFramework
 @author     wuyong
 @copyright  Copyright © 2018年 wuyong.site. All rights reserved.
 @updated    2018/9/18
 */

#import "NSFileManager+FWFramework.h"
#import <AVFoundation/AVFoundation.h>

@implementation NSFileManager (FWFramework)

#pragma mark - Path

+ (NSString *)fwPathSearch:(NSSearchPathDirectory)directory
{
    NSArray *directories = NSSearchPathForDirectoriesInDomains(directory, NSUserDomainMask, YES);
    return directories.count > 0 ? [directories objectAtIndex:0] : nil;
}

+ (NSString *)fwPathHome
{
    return NSHomeDirectory();
}

+ (NSString *)fwPathDocument
{
    return [self fwPathSearch:NSDocumentDirectory];
}

+ (NSString *)fwPathCaches
{
    return [self fwPathSearch:NSCachesDirectory];
}

+ (NSString *)fwPathLibrary
{
    return [self fwPathSearch:NSLibraryDirectory];
}

+ (NSString *)fwPathPreference
{
    return [[self fwPathLibrary] stringByAppendingPathComponent:@"Preference"];
}

+ (NSString *)fwPathTmp
{
    return NSTemporaryDirectory();
}

+ (NSString *)fwPathBundle
{
    return [[NSBundle mainBundle] bundlePath];
}

+ (NSString *)fwPathResource
{
    return [[NSBundle mainBundle] resourcePath];
}

+ (NSString *)fwAbbreviateTildePath:(NSString *)path
{
    return [path stringByAbbreviatingWithTildeInPath];
}

+ (NSString *)fwExpandTildePath:(NSString *)path
{
    return [path stringByExpandingTildeInPath];
}

#pragma mark - Size

+ (unsigned long long)fwFolderSize:(NSString *)folderPath
{
    NSArray *contents = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:folderPath error:nil];
    NSEnumerator *contentsEnumurator = [contents objectEnumerator];
    
    NSString *file;
    unsigned long long folderSize = 0;
    
    while (file = [contentsEnumurator nextObject]) {
        NSDictionary *fileAttributes = [[NSFileManager defaultManager] attributesOfItemAtPath:[folderPath stringByAppendingPathComponent:file] error:nil];
        folderSize += [[fileAttributes objectForKey:NSFileSize] intValue];
    }
    return folderSize;
}

+ (double)fwAvailableDiskSize
{
    NSDictionary *attributes = [self.defaultManager attributesOfFileSystemForPath:[self fwPathDocument] error:nil];
    return [attributes[NSFileSystemFreeSize] unsignedLongLongValue] / (double)0x100000;
}

#pragma mark - Addition

+ (BOOL)fwSkipBackup:(NSString *)path
{
    return [[NSURL.alloc initFileURLWithPath:path] setResourceValue:@(YES) forKey:NSURLIsExcludedFromBackupKey error:nil];
}

@end

File: FWFramework/Component/Foundation/Extension/NSFileManager+FWFramework.swift
//
//  NSFileManager+FWFramework.swift
//  FWFramework
//
//  Created by wuyong on 2019/6/28.
//  Copyright © 2019 wuyong.site. All rights reserved.
//

import Foundation

/// 搜索路径
///
/// - Parameter directory: 搜索目录
/// - Returns: 目标路径
public func FWPathSearch(_ directory: FileManager.SearchPathDirectory) -> String {
    return NSSearchPathForDirectoriesInDomains(directory, .userDomainMask, true)[0]
}

/// 沙盒路径，常量
public var FWPathHome: String {
    return NSHomeDirectory()
}

/// 文档路径，iTunes会同步备份
public var FWPathDocument: String {
    return FWPathSearch(.documentDirectory)
}

/// 缓存路径，系统不会删除，iTunes会删除
public var FWPathCaches: String {
    return FWPathSearch(.cachesDirectory)
}

/// Library路径
public var FWPathLibrary: String {
    return FWPathSearch(.libraryDirectory)
}

/// 配置路径，配置文件保存位置
public var FWPathPreference: String {
    return (FWPathLibrary as NSString).appendingPathComponent("Preference")
}

/// 临时路径，App退出后可能会删除
public var FWPathTmp: String {
    return NSTemporaryDirectory()
}

/// bundle路径，不可写
public var FWPathBundle: String {
    return Bundle.main.bundlePath
}

/// 资源路径，不可写
public var FWPathResource: String {
    return Bundle.main.resourcePath ?? ""
}
